### ISSUE TITLE
[java] AvoidDecimalLiteralsInBigDecimalConstructor FP with PMD 6.30.0

### ISSUE THREAD

**Issue by testation21** [pmd/pmd#3080]
**Affects PMD Version:** 6.30.0

**Rule:** AvoidDecimalLiteralsInBigDecimalConstructor
https://pmd.github.io/pmd-6.30.0/pmd_rules_java_errorprone.html#avoiddecimalliteralsinbigdecimalconstructor

**Description:** 
When call toString() method in BigDecimalConstructor, pmd shows alarms which are False Positives I think.
Please, check this case.

Thank you.

**Code Sample demonstrating the issue:**

```
import java.math.BigDecimal;
class TestAvoidDecimalLiteralsInBigDecimalConstructor{
        public static void main(String[] args){
                BigDecimal d = new BigDecimal(Double.toString(0.1));
        }
}
```

**Expected outcome:**
Don't show any alarms.

**Running PMD through:** *[CLI]*

**Comment by YoonhoChoi96** [comment]
Hi,

There are some other false positive cases: 
1) when there is only zero value below the decimal point like `1.0`,
2) when there is integer value with literal `d` or `f`,
3) when there is `Double.MAX_VALUE` or `Float.MAX_VALUE.

Code sample:
```
import java.math.BigDecimal;
class TestAvoidDecimalLiteralsInBigDecimalConstructor{
        public static void main(String[] args){
                System.out.println( new BigDecimal( 0.0 ) );    //case 1
                System.out.println( new BigDecimal( 1.0 ) );    //case 1
                System.out.println( new BigDecimal( 1d ) );     //case 2
                System.out.println( new BigDecimal( Double.MAX_VALUE ) ); //case3
        }
}
```

Then the result of running the code is
```
0
1
1
179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368
```
which is exact values.

Please, check this cases.
Thank you.

**Comment by oowekyala** [comment]
> There are some other false positive cases:

> when there is only zero value below the decimal point like 1.0,
> when there is integer value with literal d or f,
    
These are not really false positives. Adding `.0` or `d` or `f` is useless. The problem the rule flags is the usage of the constructor with a double parameter. If you were using an int literal instead, you'd be using the constructor `BigDecimal(int)`, which doesn't have the same problem.

Otherwise by that token any decimal literal which can be represented exactly by a `double` should be whitelisted by the rule, like `2.125`.

>  when there is Double.MAX_VALUE or `Float.MAX_VALUE.

Maybe this should be whitelisted by the rule.

>    BigDecimal d = new BigDecimal(Double.toString(0.1));

Similarly, `Double.toString` returns a representation of the double value, not what was written in the source code as the literal.  
If you're using Double.toString with a double literal, you should be using a string literal directly

```
jshell> Double.toString(100.0000000000000001)
$37 ==> "100.0"

jshell> new BigDecimal(Double.toString(100.0000000000000001)).equals(new BigDecimal("100.0000000000000001"))
$38 ==> false
```

**Comment by YoonhoChoi96** [comment]
Hi, thanks for your kind answer.
I have some questions about your answer.

> These are not really false positives. Adding `.0` or `d` or `f` is useless. The problem the rule flags is the usage of the constructor with a double parameter. If you were using an int literal instead, you'd be using the constructor `BigDecimal(int)`, which doesn't have the same problem.

Yes, I totally agree.

However, based on the description of the rule,
I understood the purpose of the rule is whitelisting the `decimal literal`
which affects `imprecision` of the value. 

Therefore, I thought that it is a quite different problem between
`2.125 should be changed to "2.125" for precise value` and 
`useless 0.0 should be changed to 0` or `useless 'd' or 'f' should be removed`.

What if add the description about the second case which is
`Don't use Useless literals in BigDecimal Constructor` or just
`Don't use Useless literals`?
It may be a new rule or helpful description of the rule for users.

> > when there is Double.MAX_VALUE or `Float.MAX_VALUE.
> 
> Maybe this should be whitelisted by the rule.

May I ask there is any other reason?
 
> > BigDecimal d = new BigDecimal(Double.toString(0.1));
> 
> Similarly, `Double.toString` returns a representation of the double value, not what was written in the source code as the literal.
> If you're using Double.toString with a double literal, you should be using a string literal directly
> 
> ```
> jshell> Double.toString(100.0000000000000001)
> $37 ==> "100.0"
> 
> jshell> new BigDecimal(Double.toString(100.0000000000000001)).equals(new BigDecimal("100.0000000000000001"))
> $38 ==> false
> ```

Sorry for giving you a wrong example....
I'll check more exactly and report for the next time.
I really appreciate your kind explanation.

Thank you:)

**Comment by oowekyala** [comment]
> However, based on the description of the rule,
> I understood the purpose of the rule is whitelisting the decimal literal
> which affects imprecision of the value.

> Therefore, I thought that it is a quite different problem between
> 2.125 should be changed to "2.125" for precise value and
> useless 0.0 should be changed to 0 or useless 'd' or 'f' should be removed.

> What if add the description about the second case which is
> Don't use Useless literals in BigDecimal Constructor or just
> Don't use Useless literals?
> It may be a new rule or helpful description of the rule for users.

I think you're right, and it makes sense to describe this case in the documentation. The problem here is not really loss of precision then, it's more about consistency and teaching, that it's error-prone to use that constructor. Eg if someone later changes the value without paying attention, they might end up with a loss of precision, which would not have happened if a string literal was there from the start.

It could also be nice to describe why a literal is not ok, but a double gotten from somewhere else is.

I would think a new rule to remove useless decimal suffixes could be useful (expect when it's required, eg `var d = 0.0`), as it's not limited to BigDecimal constructor. It could be use a code style rule, as it doesn't really make a difference either way. Some people may even prefer the opposite convention, where a decimal literal is used every time a double is expected, even though the language would widen an integer literal to a double.

> May I ask there is any other reason?

Do you mean, any other reason to whitelist `Double` and `Float` `MAX_VALUE` fields? Or any other constructs that should be whitelisted?

**Comment by YoonhoChoi96** [comment]
> I would think a new rule to remove useless decimal suffixes could be useful (expect when it's required, eg var d = 0.0), as it's not limited to BigDecimal constructor. 

It's great!

> Do you mean, any other reason to whitelist `Double` and `Float` `MAX_VALUE` fields? Or any other constructs that should be whitelisted?

I mean the first one, `Double.MAX_VALUE`.
Thank you.

**Comment by maikelsteneker** [comment]
After the discussion above, I'm not sure what exactly the behaviour of this rule should be, but I noticed that in PMD 6.31.0, the rule also triggers on examples where a double is passed into the BigDecimal constructor directly: `value.doubleValue()`
I can imagine similar reasoning applies here about the precision of the value, but at the very least the name of the rule suggests it applies to **literals** rather than any double value.

**Comment by YoonhoChoi96** [comment]
> After the discussion above, I'm not sure what exactly the behaviour of this rule should be, but I noticed that in PMD 6.31.0, the rule also triggers on examples where a double is passed into the BigDecimal constructor directly: `value.doubleValue()`

To the best of my knowledge, 
when using decimal literals(like 0.1, 3.5, 4.86....) in constructor of BigDecimal,
the value doesn't represent precise value of it.

For example, when I run [Example 1] code,
I expect to print the value `1.21` intuitively.
However, the result is printed like
`1.20999999999999996447286321199499070644378662109375`.
```
//Example 1
import java.math.BigDecimal;
class TestAvoidDecimalLiteralsInBigDecimalConstructor{
        public static void main(String[] args){
                Double value = 1.21;
                System.out.println( new BigDecimal( value.doubleValue()));
        }
}
```
This is the main problem.
Then, how could I get the precise value?
According to the rule, Using String literal(like "0.1", "3.5", "4.86") in constructor
could prevent this situation.
```
//Example 2
import java.math.BigDecimal;
class TestAvoidDecimalLiteralsInBigDecimalConstructor{
        public static void main(String[] args){
                Double value = 1.21;
                String valueStr = value.toString();    //make the value Double to String
                System.out.println( new BigDecimal( valueStr));   //put String literals
        }
}
```
[Example 2] is written based on the intent of the rule.
It print 1.21 precisely.

As you can see from the [Example 1], 
`value.doubleValue()` returns `double` type, which is decimal literal,
then print imprecise value. 
Therefore, `value.doubleValue()` could not be False Positive.

I wish it might be helpful to you.

**Comment by oowekyala** [comment]
>  >  > when there is Double.MAX_VALUE or `Float.MAX_VALUE.

>   > Maybe this should be whitelisted by the rule.

> May I ask there is any other reason?

They are not literals, and there's no loss of precision wen using them. I think it is a FP, I don't remember what my hesitation was...

**Comment by YoonhoChoi96** [comment]
Got it! Thanks for replying.

**Comment by jsotuyod** [comment]
I'm closing this as PMD 7.0.0 will no longer report on `new BigDecimal(Double.toString(0.1));` (which is effectively equal to `new BigDecimal("0.1");` even if more cumbersome.

A new issue can be open to whitelist `MAX_VALUE` / `MIN_VALUE` if needed.
